docs/docs/tutorial/creating-a-source-plugin/part-4/index.mdx

```diff
@@ -0,0 +1,866 @@
+- `verbose`: Print a message to the terminal that is only visible when the "verbose" flag is enabled (e.g. `gatsby build --verbose`).
+- `activityTimer`: Print an informational message to the terminal that has a timer attached to it (e.g. how long that step took).
+- `setErrorMap`: Set a custom error map to the reporter. This allows the reporter to extend Gatsby's internal error map.
+
+You'll learn how to use `panicOnBuild`, `activityTimer`, and `setErrorMap` in the next three tasks as they are the most relevant and common for plugins.
+
+### Task: Output errors
+
+As source plugins interact with third-party APIs they are inherently prone to errors (network issues, API outages, etc.). As explained in [Part 2](/docs/tutorial/creating-a-source-plugin/part-2/#fetch-data) it's good practice to gracefully handle all possible error states, as there's nothing worse than no actionable feedback at all.
+
+In Part 2 you also added the `fetchGraphQL` function that accesses the GraphQL API inside the `api` folder. When you send a malformed request to it, it can not only give back `data` but also `errors`. Use that information to output it with `panicOnBuild`.
+
+1. Go to your `plugin/src/source-nodes.ts` file. Destructure `reporter` from `gatsbyApi` at the top of `sourceNodes`. Add `errors` to the destructuring statement for the `fetchGraphQL` response. Also add a conditional statement when `errors` is truthy directly below `fetchGraphQL`:
+
+    ```ts:title=plugin/src/source-nodes.ts
+    // Imports
+
+    export const sourceNodes: GatsbyNode[`sourceNodes`] = async (gatsbyApi) => {
+      // highlight-next-line
+      const { reporter } = gatsbyApi
+
+      // Types
+
+      // highlight-next-line
+      const { data, errors } = await fetchGraphQL<IApiResponse>(
+        `http://localhost:4000/graphql`,
+        `query FetchApi {
+          posts {
+            id
+            slug
+            title
+            image {
+              url
+              alt
+              width
+              height
+            }
+            author
+          }
+          authors {
+            id
+            name
+          }
+        }
+        `
+      )
+
+      // highlight-start
+      if (errors) {
+        // Do something...
+      }
+      // highlight-end
+
+      // Rest of sourceNodes
+    }
+
+    // Rest of file
+    ```
+
+1. Inside the `errors` conditional, use `panicOnBuild` to output the error message that you're receiving from your API. `error`, `panicOnBuild`, and `panic` accept a `string`, `Error` or a structured error (you'll learn about the latter in [Task: Define custom errors](#task-define-custom-errors)).
+
+    ```ts:title=plugin/src/source-nodes.ts
+    // Imports
+
+    export const sourceNodes: GatsbyNode[`sourceNodes`] = async (gatsbyApi) => {
+      // Rest of sourceNodes
+
+      if (errors) {
+        // highlight-next-line
+        reporter.panicOnBuild(errors[0].message)
```

Comment:
should there also be "early return" here as we probably don't want to just continue trying to create nodes if there are errors reported?